File: material/code/examples/mc-pi.cpp
#include <chrono>
#include <cmath>
#include <iostream>
#include <random>


int main(int argc, char *argv[]) {
    // default parameters
    size_t nSample = 32 * 1024;  // number of sample points

    // override with command line arguments
    int i = 1;
    if (argc > i) nSample = atoi(argv[i]);
    ++i;
    
    // start measurement
    auto start = std::chrono::steady_clock::now();

    // draw nSamples points and check if they are in the target region
    size_t numHits = 0;

    // initialize a random number generator
    auto seed = 0;
    std::mt19937_64 rng(seed);
    // initialize a uniform distribution in [0 ... 1]
    std::uniform_real_distribution<double> uniform(0., 1.);

    for (size_t i = 0; i < nSample; ++i) {
        auto x = uniform(rng);
        auto y = uniform(rng);

        if (x * x + y * y <= 1.)
            ++numHits;
    }

    // finish measurement
    auto end = std::chrono::steady_clock::now();
    auto elapsedSeconds = std::chrono::duration<double>(end - start);

    std::cout << "  #samples:     " << nSample << "\n";
    std::cout << "  elapsed time: " << 1e3 * elapsedSeconds.count() << " ms\n";
    std::cout << "  MLUP/s:       " << 1e-6 * nSample / elapsedSeconds.count() << "\n";

    // compute and print final solution
    auto pi = 4. * ((double)numHits / nSample);
    std::cout << "  Estimate for pi is " << pi << std::endl;
    std::cout << "  Error is " << std::abs(pi - M_PI) / M_PI << std::endl;

    return 0;
}

File: material/code/examples/mc-pi.f90
program mc_pi
  use iso_fortran_env, only: real64
  use omp_lib
  implicit none

  integer :: n_sample, i, num_hits
  real(real64) :: x, y, pi, pi_ref, elapsed
  real(real64) :: t_start, t_end
  character(len=32) :: arg

  ! default parameters
  n_sample = 32 * 1024

  ! override with command line arguments
  if (command_argument_count() >= 1) then
    call get_command_argument(1, arg)
    read(arg, *) n_sample
  end if

  t_start = omp_get_wtime()

  num_hits = 0
  do i = 1, n_sample
    call random_number(x)
    call random_number(y)
    if (x*x + y*y <= 1.0_real64) num_hits = num_hits + 1
  end do

  t_end   = omp_get_wtime()
  elapsed = t_end - t_start

  pi     = 4.0_real64 * real(num_hits, real64) / real(n_sample, real64)
  pi_ref = 4.0_real64 * atan(1.0_real64)

  write(*, '(A,I0)')      '  #samples:        ', n_sample
  write(*, '(A,F0.3,A)') '  elapsed time:    ', 1.0e3_real64 * elapsed, ' ms'
  write(*, '(A,F0.3)')    '  MLUP/s:          ', &
    1.0e-6_real64 * real(n_sample, real64) / elapsed
  write(*, '(A,F0.6)')   '  Estimate for pi: ', pi
  write(arg, '(ES12.3)') abs(pi - pi_ref) / pi_ref
  write(*, '(A,A)')  '  Relative error:  ', trim(adjustl(arg))

end program mc_pi
